### ISSUE TITLE
Classic theme site preview: Some elements may remain interactive

### ISSUE THREAD

**Issue by t-hamano** [WordPress/gutenberg#70050]
## What problem does this address?

When the classic theme site preview is displayed, all elements inside an iframe are expected to be non-interactive via the following code:

https://github.com/WordPress/gutenberg/blob/bcc41b1d99de913f77481f718b3a3966506acce2/packages/edit-site/src/components/editor/site-preview.js#L30-L39

However, I found that this code isn't enough to make all interactive elements non-clickable or non-focusable. The scenarios I was able to find are:

### ⚒️ Summary element

- #70051

For example, you can expand or collapse the Details block content by clicking the `summary` element. The reason is that the summary element is not defined [here](https://github.com/WordPress/gutenberg/blob/bcc41b1d99de913f77481f718b3a3966506acce2/packages/dom/src/focusable.js#L31) as a focusable element.

![Image](https://github.com/user-attachments/assets/774c9e5e-1dbf-4130-86cf-df3e8597bcb5)

### Elements that are not visible on initial load

For example, Twenty Twenty-One has a mobile menu toggle button. This element isn't initially displayed, but if you narrow down the iframe width, the button is displayed, which is interactive.

I think we'll probably need to use a MutationObserver to monitor changes to the DOM tree and make newly added elements non-interactive.

![Image](https://github.com/user-attachments/assets/aa964261-24f4-464d-9a39-1f4d192f0fa7)

### Elements with event handler

Consumers may add event handlers to non-interactive elements. In this case, `focus.focusable.find()` can't detect the element.

```html
<div class="wp-block" onclick="window.alert('clicked!')">Click Me</div>
```

### Elements with the Interactivity API

Some elements may have the directives in their markup to add specific behavior the the DOM elements of the block.  For example, the image lightbox.

![Image](https://github.com/user-attachments/assets/ef2712cd-23b6-463f-9560-e466d5961ad2)

## What is your proposed solution?

We need to find an ideal approach to correctly disable all interactive elements. Additionally, there may be other cases to deal with besides the ones I have mentioned.

Not that the approach to add the `inert` attribute to the iframe body may cause an accessibility issue. See https://github.com/WordPress/gutenberg/pull/69522#issuecomment-2713017024.

**Comment by yogeshbhutkar** [comment]
Thanks for raising the issue.

While testing, I noticed that the search label was clickable. Clicking on it focused the associated input field, allowing me to type and submit a search query. This is more of a browser behaviour.

MDN Reference:
> The browser passes the focus to its associated input.

![Image](https://github.com/user-attachments/assets/81938ce4-6902-4d44-8a14-742b4bab4c4c)

> I think we'll probably need to use a MutationObserver to monitor changes to the DOM tree and make newly added elements non-interactive.

In my testing, I found that the menu within preview isn’t being inserted or removed from the DOM, it’s simply shown and hidden by a CSS rule in an external stylesheet. Because MutationObserver only fires on DOM mutations (adding/removing nodes or changing attributes), it won’t ever see a pure CSS-only toggle ( A CSSOM change ).

I believe the `pointer-events` rule is not getting applied to it because when the display is hidden, it's not focusable.
https://github.com/WordPress/gutenberg/blob/5b33c370cfbbb91d867ca4da001faf6960ee4ac8/packages/edit-site/src/components/editor/site-preview.js#L33

> Consumers may add event handlers to non-interactive elements. In this case, focus.focusable.find() can't detect the element.

Event handlers can be added in various ways, and based on my research, there's no standardized API for detecting elements with event handlers, which adds to the complexity.

I agree with not using `inert` as it would remove some elements from the Accessibility Tree, but I think applying `pointer-events: none` on the iframe's body can be a slight improvement, as that can prevent certain cases like these:

```<div class="wp-block" onclick="window.alert('clicked!')">Click Me</div>```

It would also not let the user click on the `label` and therefore, the focus will not be moved to the corresponding input.

What do you think?

**Comment by t-hamano** [comment]
@yogeshbhutkar Thanks for investigating!

> Event handlers can be added in various ways, and based on my research, there's no standardized API for detecting elements with event handlers, which adds to the complexity.

Yes, this is the most difficult problem to solve.

> I agree with not using `inert` as it would remove some elements from the Accessibility Tree, but I think applying `pointer-events: none` on the iframe's body can be a slight improvement, as that can prevent certain cases like these:
> ```
> <div class="wp-block" onclick="window.alert('clicked!')">Click Me</div>
> ```

I lean toworad to agree this approach.


@stokesman @Mamaduka @joedolson Do you have any ideas?

**Comment by joedolson** [comment]
Setting all focusable elements to have `aria-hidden="true"` is, in my opinion, the larger problem here. It's not clear from me whether this issue is exploring removing that, but it's a major accessible problem - anything marked with `aria-hidden="true"` is going to be hidden from screen reader users. 

I'll be honest: this is a known problem - almost all methods of removing interactivity also remove visibility for screen readers. We need to be able to disable interactions without changing the semantics or visibility of the element. E.g., buttons and links should still report as buttons or links, even if they are not interactive. 

I'm not totally clear what the benefit is of disabling interactivity completely; doesn't that reduce the value of the design view?

For links, you can use `<a role="link" aria-disabled="true">`, along with removing the `href` attribute to mark them as disabled while still reporting as links. Buttons can be marked as disabled in HTML; so that would be fine. 

But for things like details/summary is trickier. Adding `tabindex="-1"` so it's read without being focusable will help, and disabling `pointer-events` will stop mouse interactions; but that won't disable screen readers from accessing the interaction.

**Comment by t-hamano** [comment]
@joedolson Thanks for the feedback!

> I'm not totally clear what the benefit is of disabling interactivity completely; doesn't that reduce the value of the design view?

Perhaps we should discuss this point again. I recall there being some discussion somewhere about whether or not to expose this site preview in the first place.

Besides disabling interactivity completely, I think we could also consider one of two other approaches:

- Leave the site preview interactive as is. This could be similar to the Customizer preview.
- Don't expose the site preview. We might be able to show the StyleBook instead.

cc @WordPress/gutenberg-design

**Comment by t-hamano** [comment]
Since the 6.9 RC1 release is just around the corner, let's punt this issue to 6.9.

**Comment by t-hamano** [comment]
This issue has been punted to 6.9. Let's move this issue to the 6.9 Editor Tasks board and see what's possible in 6.9.

**Comment by t-hamano** [comment]
I'm wondering if there's a way to make the site editor interactive while preventing unintended behavior.

Perhaps we could reuse the Customizer preview implementation.

Link handling: https://github.com/WordPress/wordpress-develop/blob/1ab39ce269a0dd718f6e547162403c5eb11219ba/src/js/_enqueues/wp/customize/preview.js#L147

Form submit handling: https://github.com/WordPress/wordpress-develop/blob/1ab39ce269a0dd718f6e547162403c5eb11219ba/src/js/_enqueues/wp/customize/preview.js#L193

Determine if a link is safe to navigate: https://github.com/WordPress/wordpress-develop/blob/1ab39ce269a0dd718f6e547162403c5eb11219ba/src/js/_enqueues/wp/customize/preview.js#L283


I'd like to see if this is possible.

**Comment by t-hamano** [comment]
> I'd like to see if this is possible.

I closed #71370 to explore consistent interaction in the future. If we can't find an ideal approach for the 6.9 release, let's punt it to 7.0.
